File: src/algorithms/trees/advanced/segment-tree-range-sum/__tests__/segment-tree-range-sum.test.ts
import { describe, it, expect } from "vitest";
import { segmentTreeRangeSum } from "../sources/segment-tree-range-sum.ts?fn";

describe("segmentTreeRangeSum", () => {
  it("queries range sum for default input", () => {
    const result = segmentTreeRangeSum(
      [1, 3, 5, 7, 9, 11],
      [
        [1, 3],
        [0, 5],
      ],
    ) as number[];
    expect(result[0]).toBe(15); // 3+5+7
    expect(result[1]).toBe(36); // 1+3+5+7+9+11
  });

  it("handles single element query", () => {
    const result = segmentTreeRangeSum([4, 2, 6], [[1, 1]]) as number[];
    expect(result[0]).toBe(2);
  });

  it("handles full range query", () => {
    const arr = [1, 2, 3, 4, 5];
    const result = segmentTreeRangeSum(arr, [[0, 4]]) as number[];
    expect(result[0]).toBe(15);
  });

  it("handles multiple queries on same array", () => {
    const result = segmentTreeRangeSum(
      [10, 20, 30, 40, 50],
      [
        [0, 1],
        [2, 4],
        [1, 3],
      ],
    ) as number[];
    expect(result[0]).toBe(30);
    expect(result[1]).toBe(120);
    expect(result[2]).toBe(90);
  });
});

File: src/algorithms/trees/advanced/segment-tree-range-sum/__tests__/SegmentTreeRangeSum_test.java
// javac *.java && java -ea SegmentTreeRangeSum_test
import java.util.List;

public class SegmentTreeRangeSum_test {
    public static void main(String[] args) {
        SegmentTreeRangeSum stSum = new SegmentTreeRangeSum();

        // test: default input
        int[][] queries1 = {{1, 3}, {0, 5}};
        List<Integer> result1 = stSum.segmentTreeRangeSum(new int[]{1, 3, 5, 7, 9, 11}, queries1);
        assert result1.get(0) == 15 : "First range sum failed: " + result1.get(0);
        assert result1.get(1) == 36 : "Second range sum failed: " + result1.get(1);

        // test: single element
        int[][] queries2 = {{1, 1}};
        List<Integer> result2 = stSum.segmentTreeRangeSum(new int[]{4, 2, 6}, queries2);
        assert result2.get(0) == 2 : "Single element failed";

        // test: full range
        int[][] queries3 = {{0, 4}};
        List<Integer> result3 = stSum.segmentTreeRangeSum(new int[]{1, 2, 3, 4, 5}, queries3);
        assert result3.get(0) == 15 : "Full range failed";

        // test: multiple queries
        int[][] queries4 = {{0, 1}, {2, 4}, {1, 3}};
        List<Integer> result4 = stSum.segmentTreeRangeSum(new int[]{10, 20, 30, 40, 50}, queries4);
        assert result4.get(0) == 30 : "Multiple [0] failed";
        assert result4.get(1) == 120 : "Multiple [1] failed";
        assert result4.get(2) == 90 : "Multiple [2] failed";

        System.out.println("All tests passed!");
    }
}
